gustaf/spline/ffd.py

```diff
@@ -0,0 +1,429 @@
+
+    def insert_knots(self, parametric_dimension, knots):
+        """
+        Wrapper for Spline.insert_knots
+
+        Parameters
+        -----------
+        *args:
+        **kwargs:
+
+        Returns
+        --------
+        None
+        """
+        if "knot_vectors" in self._spline.required_properties:
+            raise NotImplementedError(
+                "Can not perform knot insertion on Bezier spline.")
+        # scale knots from the original bounds into the new bounds
+        dim_bounds = self._orig_para_dim_bounds[parametric_dimension]
+        knots = \
+            (np.array(knots)-dim_bounds[0]*1)/(dim_bounds[-1]-dim_bounds[0])
+        self._spline.insert_knots(parametric_dimension, knots)
+
+    def remove_knots(self, parametric_dimension, knots, tolerance=1e-8):
+        """
+        Wrapper for Spline.remove_knots
+
+        Parameters
+        -----------
+        *args:
+        **kwargs:
+
+        Returns
+        --------
+        None
+        """
+        if "knot_vectors" in self._spline.required_properties:
+            raise NotImplementedError(
+                "Can not perform knot insertion on Bezier spline.")
+        # scale knots from the original bounds into the new bounds
+        dim_bounds = self._orig_para_dim_bounds[parametric_dimension]
+        knots = \
+            (np.array(knots)-dim_bounds[0]*1)/(dim_bounds[-1]-dim_bounds[0])
+        self._spline.remove_knots(
+            parametric_dimension, knots, tolerance=tolerance)
+        self._is_calculated = False
+
+    def reduce_degree(self, *args, **kwargs):
+        """
+        Wrapper for Spline.reduce_degree
+
+        Parameters
+        -----------
+        *args:
+        **kwargs:
+
+        Returns
+        --------
+        None
+        """
+        self._spline.reduce_degree(*args, **kwargs)
+        self._is_calculated = False
+
+    def show(self, title: str = "Gustaf - FFD", **kwargs) -> Any:
+        """
+        Visualize. Shows the deformed mesh and the current spline. Currently 
+        visualization is limited to vedo.
+
+        Parameters
+        ----------
+        title: str
+            Title of the vedo window. Defaults to "Gustaf - FFD".
+        kwargs: Any
+            Arbitrary keyword arguments. These are passed onto the vedo
+            functions. Please be aware, that no checking of these are performed
+            in this function.
+
+        Returns
+        -------
+        Any:
+            Returns, if applicable, the vedo plotter. 'close=False' as argument
+            to get the plotter.
+        """
+        if settings.VISUALIZATION_BACKEND != "vedo":
+            raise NotImplementedError(
+                "Visualization of the FFD is not available for the chosen"
+                f"visualization framework -{settings.VISUALIZATION_BACKEND}-."
+                " Please choose vedo to visualize."
+            )
+        original_mesh = self._mesh.copy()
```

Comment:
could you add `return_discrete` and `return_showable` option and return a dict that holds corresponding objects? than you can add `FFD.showable()` that  just does `show(*args, **kwargs, return_showable=True)`